src/runtime/virtcontainers/container.go

```diff
@@ -695,6 +696,28 @@ func (c *Container) createBlockDevices(ctx context.Context) error {
 			continue
 		}
 
+		// Handle directly assigned volume. Update the mount info based on the mount info json.
+		mntInfo, e := volume.VolumeMountInfo(m.Source)
+		if e != nil && !os.IsNotExist(e) {
+			c.Logger().WithError(e).WithField("mount-source", m.Source).
+				Error("failed to parse the mount info file for a direct assigned volume")
+			continue
+		}
+
+		if mntInfo != nil {
+			// Write out sandbox info file on the mount source to allow CSI to communicate with the runtime
+			if err := volume.RecordSandboxId(c.sandboxID, m.Source); err != nil {
+				c.Logger().WithError(err).Error("error writing sandbox info")
+			}
+
+			c.mounts[i].Source = mntInfo.Device
+			c.mounts[i].Type = mntInfo.FsType
+			c.mounts[i].Options = mntInfo.Options
+			m.Source = mntInfo.Device
```

Comment:
Nit: what's the difference between `c.mounts[i]` and `m`? if they are the same, can be unified into one style.

Reply:
m is a copy of the mount object that is mutated and used to create device. But here we also want to mutate the original mount object. That's why you see this "duplicate" code. 
I can change it to something like this if it's more readable:
```
			m.Source = mntInfo.Device
			m.Type = mntInfo.FsType
			m.Options = mntInfo.Options
			c.mounts[i].Source = m.Source
			c.mounts[i].Type = m.Type
			c.mounts[i].Options = m.Options
```